File: src/main/java/maifetch/ConfigLoader.java
package maifetch;

import maifetch.json.Json;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public final class ConfigLoader {
    private static final int DEFAULT_LOGO_SIZE = 20;
    private static final int DEFAULT_SCORE_COUNT = 4;

    private ConfigLoader() {
    }

    public static Config load(String[] args) throws IOException {
        return load(args, System.getenv());
    }

    public static Config load(String[] args, Map<String, String> environment) throws IOException {
        ParsedArgs parsed = ParsedArgs.parse(args);
        File configFile = parsed.configFile == null ? defaultConfigFile() : parsed.configFile;

        MutableConfig config = new MutableConfig();
        config.logoSize = DEFAULT_LOGO_SIZE;
        config.scoreCount = DEFAULT_SCORE_COUNT;
        config.configFile = configFile;

        readConfigFile(config, configFile);
        applyEnvironment(config, environment);
        applyArguments(config, parsed);
        validate(config);

        return new Config(config.accessToken, config.configFile, config.logoSize, config.scoreCount);
    }

    static File defaultConfigFile() {
        String os = System.getProperty("os.name", "").toLowerCase();
        String home = System.getProperty("user.home", ".");
        if (os.contains("win")) {
            String appData = System.getenv("APPDATA");
            if (appData != null && !appData.trim().isEmpty()) {
                return new File(appData, "maifetch.json");
            }
            return new File(home, "AppData/Roaming/maifetch.json");
        }
        if (os.contains("mac")) {
            return new File(home, "Library/Application Support/maifetch.json");
        }
        String xdg = System.getenv("XDG_CONFIG_HOME");
        if (xdg != null && !xdg.trim().isEmpty()) {
            return new File(xdg, "maifetch.json");
        }
        return new File(home, ".config/maifetch.json");
    }

    @SuppressWarnings("unchecked")
    private static void readConfigFile(MutableConfig config, File configFile) throws IOException {
        if (configFile == null || !configFile.isFile()) {
            return;
        }
        Object parsed;
        try (Reader reader = new InputStreamReader(new FileInputStream(configFile), StandardCharsets.UTF_8)) {
            parsed = Json.parse(reader);
        }
        if (!(parsed instanceof Map)) {
            throw new IOException("config file must contain a JSON object");
        }
        Map<String, Object> json = (Map<String, Object>) parsed;
        config.accessToken = stringValue(json.get("accessToken"), config.accessToken);
        config.logoSize = intValue(json.get("logoSize"), config.logoSize);
        config.scoreCount = intValue(json.get("scoreCount"), config.scoreCount);
    }

    private static void applyEnvironment(MutableConfig config, Map<String, String> environment) {
        Map<String, String> env = environment == null ? new HashMap<String, String>() : environment;
        config.accessToken = firstNonBlank(env, config.accessToken, "MAITEA_TOKEN", "MAIFETCH_TOKEN");
        config.logoSize = firstInt(env, config.logoSize, "MAITEA_LOGO_SIZE", "MAIFETCH_LOGO_SIZE");
        config.scoreCount = firstInt(env, config.scoreCount, "MAITEA_SCORE_COUNT", "MAIFETCH_SCORE_COUNT");
        String configPath = firstNonBlank(env, null, "MAITEA_CONFIG_FILE", "MAIFETCH_CONFIG_FILE");
        if (configPath != null) {
            config.configFile = new File(configPath);
        }
    }

    private static void applyArguments(MutableConfig config, ParsedArgs parsed) {
        if (parsed.accessToken != null) {
            config.accessToken = parsed.accessToken;
        }
        if (parsed.logoSize != null) {
            config.logoSize = parsed.logoSize.intValue();
        }
        if (parsed.scoreCount != null) {
            config.scoreCount = parsed.scoreCount.intValue();
        }
        if (parsed.configFile != null) {
            config.configFile = parsed.configFile;
        }
    }

    private static void validate(MutableConfig config) {
        if (config.accessToken == null || config.accessToken.trim().isEmpty()) {
            throw new IllegalArgumentException("access token is required");
        }
        if (config.scoreCount < 1) {
            throw new IllegalArgumentException("score count must be at least 1");
        }
        if (config.scoreCount > 12) {
            throw new IllegalArgumentException("score count cannot be higher than 12");
        }
    }

    private static String firstNonBlank(Map<String, String> env, String fallback, String first, String second) {
        String value = env.get(first);
        if (value == null || value.trim().isEmpty()) {
            value = env.get(second);
        }
        return value == null || value.trim().isEmpty() ? fallback : value;
    }

    private static int firstInt(Map<String, String> env, int fallback, String first, String second) {
        String value = firstNonBlank(env, null, first, second);
        return value == null ? fallback : Integer.parseInt(value);
    }

    private static String stringValue(Object value, String fallback) {
        return value instanceof String ? (String) value : fallback;
    }

    private static int intValue(Object value, int fallback) {
        if (value instanceof Number) {
            return ((Number) value).intValue();
        }
        if (value instanceof String && !((String) value).trim().isEmpty()) {
            return Integer.parseInt((String) value);
        }
        return fallback;
    }

    private static final class MutableConfig {
        private String accessToken = "";
        private File configFile;
        private int logoSize;
        private int scoreCount;
    }

    static final class ParsedArgs {
        private String accessToken;
        private File configFile;
        private Integer logoSize;
        private Integer scoreCount;

        static ParsedArgs parse(String[] args) {
            ParsedArgs parsed = new ParsedArgs();
            for (int i = 0; i < args.length; i++) {
                String arg = args[i];
                if ("--help".equals(arg) || "-h".equals(arg)) {
                    throw new HelpRequested();
                }
                if (arg.startsWith("--access-token=")) {
                    parsed.accessToken = arg.substring("--access-token=".length());
                } else if ("--access-token".equals(arg) || "-a".equals(arg) || "-t".equals(arg)) {
                    parsed.accessToken = requireValue(args, ++i, arg);
                } else if (arg.startsWith("--logo-size=")) {
                    parsed.logoSize = Integer.valueOf(arg.substring("--logo-size=".length()));
                } else if ("--logo-size".equals(arg) || "-l".equals(arg)) {
                    parsed.logoSize = Integer.valueOf(requireValue(args, ++i, arg));
                } else if (arg.startsWith("--score-count=")) {
                    parsed.scoreCount = Integer.valueOf(arg.substring("--score-count=".length()));
                } else if ("--score-count".equals(arg) || "-s".equals(arg)) {
                    parsed.scoreCount = Integer.valueOf(requireValue(args, ++i, arg));
                } else if (arg.startsWith("--config-file=")) {
                    parsed.configFile = new File(arg.substring("--config-file=".length()));
                } else if ("--config-file".equals(arg) || "-c".equals(arg)) {
                    parsed.configFile = new File(requireValue(args, ++i, arg));
                } else {
                    throw new IllegalArgumentException("unknown argument: " + arg);
                }
            }
            return parsed;
        }

        private static String requireValue(String[] args, int index, String flag) {
            if (index >= args.length) {
                throw new IllegalArgumentException(flag + " requires a value");
            }
            return args[index];
        }
    }

    public static final class HelpRequested extends RuntimeException {
        private static final long serialVersionUID = 1L;
    }
}

File: src/main/kotlin/maifetch/ConfigLoader.kt
package maifetch

import maifetch.json.Json
import java.io.File
import java.nio.charset.StandardCharsets

object ConfigLoader {
    private const val DEFAULT_LOGO_SIZE = 20
    private const val DEFAULT_SCORE_COUNT = 4

    fun load(args: Array<String>): Config = load(args, System.getenv())

    fun load(args: Array<String>, environment: Map<String, String>): Config {
        val parsed = ParsedArgs.parse(args)
        val mutable = MutableConfig(
            configFile = parsed.configFile ?: defaultConfigFile(),
            logoSize = DEFAULT_LOGO_SIZE,
            scoreCount = DEFAULT_SCORE_COUNT,
        )

        readConfigFile(mutable)
        applyEnvironment(mutable, environment)
        applyArguments(mutable, parsed)
        validate(mutable)

        return Config(
            accessToken = mutable.accessToken.orEmpty(),
            configFile = mutable.configFile,
            logoSize = mutable.logoSize,
            scoreCount = mutable.scoreCount,
        )
    }

    fun defaultConfigFile(): File {
        val os = System.getProperty("os.name", "").lowercase()
        val home = System.getProperty("user.home", ".")
        if ("win" in os) {
            val appData = System.getenv("APPDATA")
            return if (!appData.isNullOrBlank()) File(appData, "maifetch.json") else File(home, "AppData/Roaming/maifetch.json")
        }
        if ("mac" in os) {
            return File(home, "Library/Application Support/maifetch.json")
        }
        val xdg = System.getenv("XDG_CONFIG_HOME")
        return if (!xdg.isNullOrBlank()) File(xdg, "maifetch.json") else File(home, ".config/maifetch.json")
    }

    @Suppress("UNCHECKED_CAST")
    private fun readConfigFile(config: MutableConfig) {
        val file = config.configFile ?: return
        if (!file.isFile) return
        val parsed = file.inputStream().reader(StandardCharsets.UTF_8).use { Json.parse(it) }
        val json = parsed as? Map<String, Any?> ?: throw java.io.IOException("config file must contain a JSON object")
        config.accessToken = stringValue(json["accessToken"], config.accessToken)
        config.logoSize = intValue(json["logoSize"], config.logoSize)
        config.scoreCount = intValue(json["scoreCount"], config.scoreCount)
    }

    private fun applyEnvironment(config: MutableConfig, environment: Map<String, String>?) {
        val env = environment.orEmpty()
        config.accessToken = firstNonBlank(env, config.accessToken, "MAITEA_TOKEN", "MAIFETCH_TOKEN")
        config.logoSize = firstInt(env, config.logoSize, "MAITEA_LOGO_SIZE", "MAIFETCH_LOGO_SIZE")
        config.scoreCount = firstInt(env, config.scoreCount, "MAITEA_SCORE_COUNT", "MAIFETCH_SCORE_COUNT")
        firstNonBlank(env, null, "MAITEA_CONFIG_FILE", "MAIFETCH_CONFIG_FILE")?.let {
            config.configFile = File(it)
        }
    }

    private fun applyArguments(config: MutableConfig, parsed: ParsedArgs) {
        parsed.accessToken?.let { config.accessToken = it }
        parsed.logoSize?.let { config.logoSize = it }
        parsed.scoreCount?.let { config.scoreCount = it }
        parsed.configFile?.let { config.configFile = it }
    }

    private fun validate(config: MutableConfig) {
        require(!config.accessToken.isNullOrBlank()) { "access token is required" }
        require(config.scoreCount >= 1) { "score count must be at least 1" }
        require(config.scoreCount <= 12) { "score count cannot be higher than 12" }
    }

    private fun firstNonBlank(env: Map<String, String>, fallback: String?, first: String, second: String): String? {
        val value = env[first].takeUnless { it.isNullOrBlank() } ?: env[second].takeUnless { it.isNullOrBlank() }
        return value ?: fallback
    }

    private fun firstInt(env: Map<String, String>, fallback: Int, first: String, second: String): Int =
        firstNonBlank(env, null, first, second)?.toInt() ?: fallback

    private fun stringValue(value: Any?, fallback: String?): String? = (value as? String) ?: fallback

    private fun intValue(value: Any?, fallback: Int): Int =
        when (value) {
            is Number -> value.toInt()
            is String -> value.takeIf { it.isNotBlank() }?.toInt() ?: fallback
            else -> fallback
        }

    private data class MutableConfig(
        var accessToken: String? = "",
        var configFile: File? = null,
        var logoSize: Int = 0,
        var scoreCount: Int = 0,
    )

    data class ParsedArgs(
        var accessToken: String? = null,
        var configFile: File? = null,
        var logoSize: Int? = null,
        var scoreCount: Int? = null,
    ) {
        companion object {
            fun parse(args: Array<String>): ParsedArgs {
                val parsed = ParsedArgs()
                var index = 0
                while (index < args.size) {
                    val arg = args[index]
                    when {
                        arg == "--help" || arg == "-h" -> throw HelpRequested()
                        arg.startsWith("--access-token=") -> parsed.accessToken = arg.substringAfter("=")
                        arg == "--access-token" || arg == "-a" || arg == "-t" -> parsed.accessToken = requireValue(args, ++index, arg)
                        arg.startsWith("--logo-size=") -> parsed.logoSize = arg.substringAfter("=").toInt()
                        arg == "--logo-size" || arg == "-l" -> parsed.logoSize = requireValue(args, ++index, arg).toInt()
                        arg.startsWith("--score-count=") -> parsed.scoreCount = arg.substringAfter("=").toInt()
                        arg == "--score-count" || arg == "-s" -> parsed.scoreCount = requireValue(args, ++index, arg).toInt()
                        arg.startsWith("--config-file=") -> parsed.configFile = File(arg.substringAfter("="))
                        arg == "--config-file" || arg == "-c" -> parsed.configFile = File(requireValue(args, ++index, arg))
                        else -> throw IllegalArgumentException("unknown argument: $arg")
                    }
                    index++
                }
                return parsed
            }

            private fun requireValue(args: Array<String>, index: Int, flag: String): String {
                if (index >= args.size) {
                    throw IllegalArgumentException("$flag requires a value")
                }
                return args[index]
            }
        }
    }

    class HelpRequested : RuntimeException()
}
